Support arbitrary properties in principal

GNSSYSTEMS-tg opened the pull request:
* added `data` field for `Principal`

This makes the principal more generic. This will be the last PR for `2.0`.

IsNull commented:
The idea around the principal object is that it defines a clear, abstract API to interact with a principal. Instead of such a generic field, we probably should just allow access to the data behind it.

```
const data = principal.getProperty("mycustom")
```

The implementation could either redirect the property key to the claims, or we just read all properties from the JWT claims and store them in a map inside the principal. API wise, we wont expose the JWT.

If I may ask, what kind of property do you need? Maybe it would  be reasonable to have it as first class citizen anyway?

GNSSYSTEMS-tg commented:
includes #17

GNSSYSTEMS-tg commented:
I'm not sure yet, which information will finally make it the jwt. Our use case is the authentication via an Active Directory, so multiple additional information is available (e.g. full name, email address, office). I'd prefer to allow generic fields besides the standard. Your approach looks accurate, though.

GNSSYSTEMS-tg commented:
* added `Principal.hasProperty(propertyName: string): any`
* added `Principal.getProperty(propertyName: string): any`
* added dependency to `lodash` (not yet documented)

IsNull commented on src/auth/principal.ts:
```diff
@@ -15,20 +15,24 @@ export class Principal {
 
     private _realm: string;
 
+    private _data: any;
```
Please use a Map here. Probably `Map<string, string> properties`.

IsNull commented on src/auth/services/token.service.ts:
```diff
@@ -257,7 +258,11 @@ export class TokenService {
+            _.omit(
+                jwt,
+                _.keys(this.config).filter(key => key.startsWith('jwt')),
+            ),
```
I would prefer if we hold the number of dependencies down to a bare minimum in a base library as this. That being said, I don't think we have to necessary omit the already mapped fields.

By using TS features like `Object.keys(obj)` it should be easily possible without additional dependencies. :)

IsNull commented on README.md:
```diff
@@ -271,6 +271,7 @@ export class JwtAuthConfig {
         issuerField?: string;
         audienceField?: string;
         realmField?: string;
+        dataField?: string;
```
This can be remove I guess.

GNSSYSTEMS-tg commented:
The JWT content is now accessible as whole, although I have to admit a reduced instance would be more handy to force the developer to stick to the standard.

IsNull commented on src/auth/services/token.service.ts:
```diff
@@ -257,7 +258,8 @@ export class TokenService {
             jwt[this.config.jwtAudienceField],
-            jwt[this.config.jwtRealmField]
+            jwt[this.config.jwtRealmField],
+            jwt,
```
Assigning an object directly to a map seems a bit of a stretch? Does this even compile on your end?

GNSSYSTEMS-tg replied:
I have to admit, I'm not quite sure, how the final solution is supposed to look like. As you described previously, having the complete jwt available via `getProperty` is desired modification.

Unfortunately I'm out of spare time for this. I'd like to have this feature, but I can't seem to find a proper solution in a short amount of time.

IsNull replied:
No problem, I try to finish it if I find some time. 

>how the final solution is supposed to look like

Well its basically just enumerating all properties of the object and putting them into the map.
